File: compactor.py
"""Context compaction for SkillOS — ported from claw-code's compact.rs pattern.

Prevents unbounded message growth by summarizing older messages
when token estimates exceed a threshold, while preserving recent context.
Supports optional LLM-powered summarization via async compact_messages_async().
"""

from __future__ import annotations

import asyncio
import warnings
from dataclasses import dataclass
from typing import Awaitable, Callable

# ── Model context-window registry ────────────────────────────────
MODEL_CONTEXT_WINDOWS: dict[str, int] = {
    # Qwen via OpenRouter
    "qwen/qwen3.6-plus:free": 32_000,
    "qwen/qwen3-plus": 32_000,
    # Gemini
    "gemini-2.5-flash": 1_048_576,
    "gemini-2.5-pro": 1_048_576,
    "gemini-2.0-flash": 1_048_576,
    # Gemma 4 via Ollama
    "gemma4": 128_000,
    "gemma4:e2b": 128_000,
    "gemma4:e4b": 128_000,
    "gemma4:26b": 256_000,
    "gemma4:31b": 256_000,
    # Gemma 4 via OpenRouter
    "google/gemma-4-26b-a4b-it": 131_072,
}

DEFAULT_COMPACTION_RATIO = 0.70   # compact at 70 % of context window
MIN_COMPACTION_THRESHOLD = 8_000  # never compact below this floor


@dataclass
class CompactionConfig:
    """Controls when and how message lists are compacted."""
    preserve_recent_messages: int = 4
    max_estimated_tokens: int = 10_000

    def configure_for_model(self, model_name: str) -> None:
        """Set *max_estimated_tokens* from the model's context window.

        Unknown models keep the current default.
        """
        window = MODEL_CONTEXT_WINDOWS.get(model_name)
        if window is not None:
            self.max_estimated_tokens = max(
                MIN_COMPACTION_THRESHOLD,
                int(window * DEFAULT_COMPACTION_RATIO),
            )
        # else: keep the default 10_000


def estimate_tokens(messages: list[dict]) -> int:
    """Estimate token count using char/4 heuristic (same as claw-code)."""
    total = 0
    for msg in messages:
        content = msg.get("content", "")
        if isinstance(content, str):
            total += len(content) // 4 + 1
        elif isinstance(content, list):
            # Handle multi-part content (vision messages, etc.)
            for part in content:
                if isinstance(part, dict):
                    total += len(str(part.get("text", ""))) // 4 + 1
                else:
                    total += len(str(part)) // 4 + 1
    return total


def should_compact(messages: list[dict], config: CompactionConfig) -> bool:
    """Check if messages exceed thresholds and need compaction."""
    return (
        len(messages) > config.preserve_recent_messages
        and estimate_tokens(messages) >= config.max_estimated_tokens
    )


def compact_messages(
    messages: list[dict], config: CompactionConfig
) -> tuple[list[dict], str]:
    """Compact older messages into a summary, preserving recent ones.

    Returns (compacted_messages, summary_text).
    If compaction is not needed, returns the original messages unchanged.
    """
    if messages and messages[0].get("role") == "system":
        warnings.warn(
            "compact_messages received a system prompt as messages[0]. "
            "The system prompt should be prepended at call time, not stored "
            "in the conversation list — compaction will erase it.",
            stacklevel=2,
        )

    if not should_compact(messages, config):
        return messages, ""

    keep_from = max(0, len(messages) - config.preserve_recent_messages)
    removed = messages[:keep_from]
    preserved = messages[keep_from:]

    summary = _summarize_messages(removed)
    continuation = (
        "This session continues from an earlier conversation. "
        f"Summary of prior context:\n\n{summary}\n\n"
        "Recent messages are preserved verbatim. "
        "Continue without recapping."
    )

    compacted = [{"role": "system", "content": continuation}] + preserved
    return compacted, summary


def _summarize_messages(messages: list[dict]) -> str:
    """Create a concise textual summary of a message list."""
    lines = []
    for msg in messages:
        role = msg.get("role", "unknown")
        content = msg.get("content", "")
        if isinstance(content, str):
            text = content[:160]
            if len(content) > 160:
                text += "..."
        else:
            text = str(content)[:160] + "..."
        lines.append(f"- {role}: {text}")
    return "\n".join(lines)


async def _summarize_with_llm(
    messages: list[dict],
    llm_fn: Callable[[str, str], Awaitable[str]],
) -> str:
    """Summarize using LLM. llm_fn: async (prompt, system) -> str."""
    transcript = "\n".join(
        f"[{m.get('role')}]: {str(m.get('content', ''))[:500]}"
        for m in messages
    )
    prompt = (
        "Summarize this conversation into 3-5 bullet points. "
        "Preserve key decisions, tool results, errors encountered.\n\n"
        f"Conversation ({len(messages)} messages):\n{transcript}"
    )
    try:
        return (await llm_fn(prompt, "Output only bullet points.")).strip()
    except Exception:
        return _summarize_messages(messages)  # fallback


async def compact_messages_async(
    messages: list[dict],
    config: CompactionConfig,
    llm_fn: Callable[[str, str], Awaitable[str]] | None = None,
) -> tuple[list[dict], str]:
    """Async compact with optional LLM summarization.

    If llm_fn is provided and there are enough messages, uses LLM for
    semantic summarization. Otherwise falls back to text truncation.
    """
    if messages and messages[0].get("role") == "system":
        warnings.warn(
            "compact_messages_async received a system prompt as messages[0]. "
            "The system prompt should be prepended at call time, not stored "
            "in the conversation list — compaction will erase it.",
            stacklevel=2,
        )

    if not should_compact(messages, config):
        return messages, ""

    keep_from = max(0, len(messages) - config.preserve_recent_messages)
    removed = messages[:keep_from]
    preserved = messages[keep_from:]

    if llm_fn and len(removed) >= 4:
        summary = await _summarize_with_llm(removed, llm_fn)
    else:
        summary = _summarize_messages(removed)

    continuation = [{
        "role": "system",
        "content": (
            "Session continues. Prior context:\n\n"
            f"{summary}\n\n"
            "Recent messages preserved. Continue without recapping."
        ),
    }]
    return continuation + preserved, summary

File: mobile/src/lib/llm/compactor.ts
/**
 * Context compaction — TS port of
 * C:\evolvingagents\skillos\compactor.py.
 *
 * When a `runGoal` loop's conversation grows past a token-estimated
 * threshold, compact older messages into a single summary user-message and
 * preserve the recent N turns verbatim. Two modes:
 *   - "fifo": drop old tool-result turns (v0 behavior, kept as safety net)
 *   - "llm": semantic summary via the LLM (new in M16)
 *
 * The same char/4 token heuristic as the Python runtime.
 */

import type { ChatMessage } from "./client";
import type { LLMProvider } from "./provider";

export const MODEL_CONTEXT_WINDOWS: Record<string, number> = {
  "qwen/qwen3.6-plus:free": 32_000,
  "qwen/qwen3-plus": 32_000,
  "gemini-2.5-flash": 1_048_576,
  "gemini-2.5-pro": 1_048_576,
  "gemini-2.0-flash": 1_048_576,
  gemma4: 128_000,
  "gemma4:e2b": 128_000,
  "gemma4:e4b": 128_000,
  "gemma4:26b": 256_000,
  "gemma4:31b": 256_000,
  "google/gemma-4-26b-a4b-it": 131_072,
  // M9 local models
  "qwen2.5-1.5b-instruct-q4_k_m": 32_768,
  "gemma-2-2b-it-q4_k_m": 8_192,
  "gemma-2-2b-it-litertlm": 8_192,
};

export const DEFAULT_COMPACTION_RATIO = 0.7;
export const MIN_COMPACTION_THRESHOLD = 8_000;

export interface CompactionConfig {
  preserveRecentMessages: number;
  maxEstimatedTokens: number;
  /**
   * Minimum removed-message count for LLM summarization to engage. Below
   * this we fall through to the textual summary.
   */
  llmSummaryMinMessages: number;
}

export function defaultCompactionConfig(): CompactionConfig {
  return {
    preserveRecentMessages: 4,
    maxEstimatedTokens: 10_000,
    llmSummaryMinMessages: 4,
  };
}

export function configureForModel(
  config: CompactionConfig,
  modelName: string,
): CompactionConfig {
  const window = MODEL_CONTEXT_WINDOWS[modelName];
  if (window !== undefined) {
    return {
      ...config,
      maxEstimatedTokens: Math.max(
        MIN_COMPACTION_THRESHOLD,
        Math.floor(window * DEFAULT_COMPACTION_RATIO),
      ),
    };
  }
  return config;
}

export function estimateTokens(messages: ChatMessage[]): number {
  let total = 0;
  for (const m of messages) {
    const c = m.content ?? "";
    total += Math.floor(String(c).length / 4) + 1;
  }
  return total;
}

export function shouldCompact(
  messages: ChatMessage[],
  config: CompactionConfig,
): boolean {
  return (
    messages.length > config.preserveRecentMessages &&
    estimateTokens(messages) >= config.maxEstimatedTokens
  );
}

/** Textual fallback — just a bulleted list of short role:content previews. */
function summarizeTextual(messages: ChatMessage[]): string {
  return messages
    .map((m) => {
      const role = m.role ?? "unknown";
      const content = String(m.content ?? "");
      const short = content.length > 160 ? `${content.slice(0, 160)}…` : content;
      return `- ${role}: ${short}`;
    })
    .join("\n");
}

async function summarizeWithLLM(
  messages: ChatMessage[],
  llm: LLMProvider,
): Promise<string> {
  const transcript = messages
    .map((m) => `[${m.role}]: ${String(m.content ?? "").slice(0, 500)}`)
    .join("\n");
  const prompt =
    "Summarize this conversation into 3–5 bullet points. " +
    "Preserve concrete identifiers, numbers, file paths, tool results, errors encountered.\n\n" +
    `Conversation (${messages.length} messages):\n${transcript}`;
  try {
    const r = await llm.chat(
      [
        { role: "system", content: "Output only bullet points." },
        { role: "user", content: prompt },
      ],
      { stream: false, temperature: 0.1, maxTokens: 512 },
    );
    return r.content.trim() || summarizeTextual(messages);
  } catch {
    return summarizeTextual(messages);
  }
}

export interface CompactResult {
  messages: ChatMessage[];
  summary: string;
  /** How many messages were rolled into the summary. */
  compacted: number;
}

/**
 * Compact older messages into a summary user-message while preserving the
 * most recent N. Pure synchronous fallback (no LLM call).
 */
export function compactMessages(
  messages: ChatMessage[],
  config: CompactionConfig,
): CompactResult {
  if (!shouldCompact(messages, config)) {
    return { messages, summary: "", compacted: 0 };
  }
  const keepFrom = Math.max(0, messages.length - config.preserveRecentMessages);
  const removed = messages.slice(0, keepFrom);
  const preserved = messages.slice(keepFrom);
  const summary = summarizeTextual(removed);
  return {
    messages: [summaryMessage(summary), ...preserved],
    summary,
    compacted: removed.length,
  };
}

/** Async LLM-powered compaction. Fires through `llm.chat`. */
export async function compactMessagesAsync(
  messages: ChatMessage[],
  config: CompactionConfig,
  llm?: LLMProvider,
): Promise<CompactResult> {
  if (!shouldCompact(messages, config)) {
    return { messages, summary: "", compacted: 0 };
  }
  const keepFrom = Math.max(0, messages.length - config.preserveRecentMessages);
  const removed = messages.slice(0, keepFrom);
  const preserved = messages.slice(keepFrom);
  const summary =
    llm && removed.length >= config.llmSummaryMinMessages
      ? await summarizeWithLLM(removed, llm)
      : summarizeTextual(removed);
  return {
    messages: [summaryMessage(summary), ...preserved],
    summary,
    compacted: removed.length,
  };
}

function summaryMessage(summary: string): ChatMessage {
  return {
    role: "user",
    content:
      "Session continues. Prior context:\n\n" +
      `${summary}\n\n` +
      "Recent messages preserved verbatim. Continue without recapping.",
  };
}
